{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1:  Controls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function \"compara_campos\" returns True if all lines of two files are in the same sequence (AGI Locus Code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "\n",
    "def compara_campos(file_Germ, file_Loc_G, campo): \n",
    "    Germplasmfile = open(file_Germ)           # Open Germplasm.tsv\n",
    "    with open(file_Loc_G) as Locusgenfile:    # Open LocusGene.tsv\n",
    "        contenido_Germplasmfile = list(csv.DictReader(Germplasmfile, delimiter=\"\\t\")) # Create lists with Germplasm.tsv content\n",
    "        contenido_Locusgenfile = list(csv.DictReader(Locusgenfile, delimiter=\"\\t\"))   # Create lists with LocusGene.tsv content\n",
    "        Germplasmfile.close()    # Close Germplasm.tsv\n",
    "        Locusgenfile.close()     # Close LocusGene.tsv\n",
    "        \n",
    "        for item in range(len(contenido_Germplasmfile)): #Loop through all lines of both files\n",
    "            if contenido_Germplasmfile[item][campo] != contenido_Locusgenfile[item][campo]: # Compare each line Locus\n",
    "                return(False)   # If one line not in the same sequence -> Return False\n",
    "            return(True)        # If all lines in the same sequence -> Return True\n",
    "        \n",
    "compara_campos(\"Germplasm.tsv\", \"LocusGene.tsv\", \"Locus\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "6 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>exam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>project</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('exam',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('project',),\n",
       " ('sys',)]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql drop database exam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "6 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>exam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>project</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('exam',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('project',),\n",
       " ('sys',)]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql create database exam;\n",
    "%sql show databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use exam;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql drop table germplasm;\n",
    "%sql drop table gene;\n",
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(20) NOT NULL, germplasm_code VARCHAR(50) NOT NULL, phenotype VARCHAR(7000) NOT NULL, pubmed VARCHAR(20) NOT NULL);\n",
    "%sql DESCRIBE germplasm;\n",
    "%sql CREATE TABLE gene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, germplasm_id INTEGER NOT NULL, locus VARCHAR(20) NOT NULL, gene_code VARCHAR(15) NOT NULL, protein_length INTEGER NOT NULL);\n",
    "#%sql DESCRIBE gene;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql DESCRIBE germplasm;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function \"inserta_datos\" takes the content of echa line of boths files and insert into the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "def inserta_datos(germ_content, locgen_content):\n",
    "\n",
    "    connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "    connection.autocommit = False\n",
    "\n",
    "    try:\n",
    "        with connection.cursor() as cursor:\n",
    "        \n",
    "            sql = \"\"\"INSERT INTO germplasm (locus, germplasm_code, phenotype, pubmed) \n",
    "            VALUES ('{}', '{}', '{}', '{}')\"\"\".format(germ_content['Locus'],\n",
    "                                                      germ_content['germplasm'],\n",
    "                                                      germ_content['phenotype'],\n",
    "                                                      germ_content['pubmed'])\n",
    "            cursor.execute(sql)\n",
    "            sql = \"SELECT last_insert_id()\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()   # traertodos\n",
    "            germplasmid = results[0]['last_insert_id()']\n",
    "            \n",
    "            \n",
    "            sql = \"\"\"INSERT INTO gene (germplasm_id, locus, gene_code, protein_length) \n",
    "            VALUES ('{}', '{}', '{}', '{}')\"\"\".format(germplasmid,\n",
    "                                                      locgen_content['Locus'],\n",
    "                                                      locgen_content['Gene'],\n",
    "                                                      locgen_content['ProteinLength'])\n",
    "            cursor.execute(sql)\n",
    "            connection.commit()   \n",
    "        \n",
    "    finally:\n",
    "        print(\"\")\n",
    "        connection.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import pymysql.cursors\n",
    "\n",
    "Germfile = open(\"Germplasm.tsv\")\n",
    "LocGenfile = open(\"LocusGene.tsv\")\n",
    "germ_content = list(csv.DictReader(Germfile, delimiter=\"\\t\"))\n",
    "locgen_content = list(csv.DictReader(LocGenfile, delimiter=\"\\t\"))\n",
    "\n",
    "for item in range(len(germ_content)):\n",
    "    inserta_datos(germ_content[item], locgen_content[item])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql SELECT * FROM germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql SELECT * FROM gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "def crea_reportes(query):\n",
    "\n",
    "    connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "    connection.autocommit = False\n",
    "\n",
    "    try:\n",
    "        with connection.cursor() as cursor:\n",
    "        \n",
    "            sql = \"{}\".format(query)\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()  \n",
    "            \n",
    "            connection.commit()   \n",
    "        \n",
    "    finally:\n",
    "        print(\"\")\n",
    "        connection.close() \n",
    "        \n",
    "    return(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>locus</th>\n",
       "        <th>germplasm_code</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "        <th>id_1</th>\n",
       "        <th>germplasm_id</th>\n",
       "        <th>locus_1</th>\n",
       "        <th>gene_code</th>\n",
       "        <th>protein_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "        <td>1</td>\n",
       "        <td>1</td>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "        <td>2</td>\n",
       "        <td>2</td>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>3</td>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "        <td>3</td>\n",
       "        <td>3</td>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "        <td>4</td>\n",
       "        <td>4</td>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>5</td>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "        <td>5</td>\n",
       "        <td>5</td>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>6</td>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "        <td>6</td>\n",
       "        <td>6</td>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>7</td>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "        <td>7</td>\n",
       "        <td>7</td>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>8</td>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "        <td>8</td>\n",
       "        <td>8</td>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>9</td>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "        <td>9</td>\n",
       "        <td>9</td>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>10</td>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "        <td>10</td>\n",
       "        <td>10</td>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>11</td>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "        <td>11</td>\n",
       "        <td>11</td>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>12</td>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "        <td>12</td>\n",
       "        <td>12</td>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>13</td>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "        <td>13</td>\n",
       "        <td>13</td>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>14</td>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "        <td>14</td>\n",
       "        <td>14</td>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>15</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "        <td>15</td>\n",
       "        <td>15</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>16</td>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "        <td>16</td>\n",
       "        <td>16</td>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>17</td>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "        <td>17</td>\n",
       "        <td>17</td>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>18</td>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "        <td>18</td>\n",
       "        <td>18</td>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>19</td>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "        <td>19</td>\n",
       "        <td>19</td>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>20</td>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "        <td>20</td>\n",
       "        <td>20</td>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>21</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "        <td>21</td>\n",
       "        <td>21</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>22</td>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "        <td>22</td>\n",
       "        <td>22</td>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>23</td>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "        <td>23</td>\n",
       "        <td>23</td>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>24</td>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "        <td>24</td>\n",
       "        <td>24</td>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>25</td>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "        <td>25</td>\n",
       "        <td>25</td>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>26</td>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "        <td>26</td>\n",
       "        <td>26</td>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>27</td>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "        <td>27</td>\n",
       "        <td>27</td>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>28</td>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "        <td>28</td>\n",
       "        <td>28</td>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>29</td>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "        <td>29</td>\n",
       "        <td>29</td>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>30</td>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "        <td>30</td>\n",
       "        <td>30</td>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>31</td>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "        <td>31</td>\n",
       "        <td>31</td>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>32</td>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "        <td>32</td>\n",
       "        <td>32</td>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1, 'AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', '17369351', 1, 1, 'AT1G01040', 'DCL1', 332),\n",
       " (2, 'AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', '16891401', 2, 2, 'AT1G01060', 'LHY', 290),\n",
       " (3, 'AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', '17486125', 3, 3, 'AT1G01140', 'CIPK9', 223),\n",
       " (4, 'AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', '18199744', 4, 4, 'AT1G01220', 'FKGP', 190),\n",
       " (5, 'AT2G03720', 'SALK_042433', 'Multiple straight hairs', '16367956', 5, 5, 'AT2G03720', 'MRH6', 189),\n",
       " (6, 'AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', '15215505', 6, 6, 'AT2G03800', 'GEK1', 196),\n",
       " (7, 'AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', '17933900', 7, 7, 'AT2G04240', 'XERICO', 256),\n",
       " (8, 'AT2G05210', 'pot1-1', 'No visible phenotype.', '17627276', 8, 8, 'AT2G05210', 'POT1A', 221),\n",
       " (9, 'AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', '17419837', 9, 9, 'AT3G02130', 'RPK2', 284),\n",
       " (10, 'AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', '18484180', 10, 10, 'AT3G02140', 'TMAC2', 300),\n",
       " (11, 'AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', '21478444', 11, 11, 'AT3G02230', 'RGP1', 301),\n",
       " (12, 'AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', '17586653', 12, 12, 'AT3G02260', 'BIG', 279),\n",
       " (13, 'AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', '10821278', 13, 13, 'AT3G02310', 'SEP2', 175),\n",
       " (14, 'AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', '17672843', 14, 14, 'AT3G02680', 'NBS1', 190),\n",
       " (15, 'AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', '17568770', 15, 15, 'AT3G02850', 'SKOR', 234),\n",
       " (16, 'AT3G02870', 'vtc4-1', 'ascorbate deficient', '16595667', 16, 16, 'AT3G02870', 'VTC4', 311),\n",
       " (17, 'AT3G03260', 'hdg8-1', 'No visible phenotype.', '16778018', 17, 17, 'AT3G03260', 'HDG8', 194),\n",
       " (18, 'AT4G14790', 'pdd17', 'Defective pollen development.', '19237690', 18, 18, 'AT4G14790', 'SUV3', 312),\n",
       " (19, 'AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', '16297066', 19, 19, 'AT4G15210', 'BAM5', 313),\n",
       " (20, 'AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', '10982425', 20, 20, 'AT4G15560', 'DXS', 219),\n",
       " (21, 'AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', '18772186', 21, 21, 'AT4G15570', 'MAA3', 294),\n",
       " (22, 'AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', '20388662', 22, 22, 'AT4G15802', 'HSBP', 254),\n",
       " (23, 'AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', '17513499', 23, 23, 'AT4G15880', 'ESD4', 265),\n",
       " (24, 'AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', '12747832', 24, 24, 'AT4G16420', 'ADA2B', 279),\n",
       " (25, 'AT4G16480', 'atint4-2', 'No visible phenotype.', '16603666', 25, 25, 'AT4G16480', 'INT4', 284),\n",
       " (26, 'AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', '18799749', 26, 26, 'AT5G10480', 'PAS2', 301),\n",
       " (27, 'AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', '17960244', 27, 27, 'AT5G10510', 'AIL6', 310),\n",
       " (28, 'AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', '18779216', 28, 28, 'AT5G11110', 'SPS2', 232),\n",
       " (29, 'AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', '16891401', 29, 29, 'AT5G11260', 'HY5', 221),\n",
       " (30, 'AT5G11510', 'myb3r4-1', 'No visible phenotype.', '17287251', 30, 30, 'AT5G11510', 'MYB3R-4', 336),\n",
       " (31, 'AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', '19413687', 31, 31, 'AT5G12200', 'PYD2', 310),\n",
       " (32, 'AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', '12345678', 32, 32, 'AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 270,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasm, gene  WHERE germplasm.id = gene.germplasm_id;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "Exercise1 = crea_reportes(\"SELECT * FROM germplasm, gene  WHERE germplasm.id = gene.germplasm_id\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Create a joined report that only includes the Genes SKOR and MAA3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM germplasm INNER JOIN gene ON \\\n",
    "     germplasm.id = gene.germplasm_id AND \\\n",
    "     gene.gene_code = 'SKOR' OR \\\n",
    "     germplasm.id = gene.germplasm_id AND \\\n",
    "     gene.gene_code = 'MAA3'; "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "Exercise2 = crea_reportes(\"\"\"SELECT * FROM germplasm INNER JOIN gene ON \\\n",
    "     germplasm.id = gene.germplasm_id AND \\\n",
    "     gene.gene_code = 'SKOR' OR \\\n",
    "     germplasm.id = gene.germplasm_id AND \\\n",
    "     gene.gene_code = 'MAA3'\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Number Of Matches</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(4,)]"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number Of Matches\" FROM germplasm WHERE locus REGEXP 'AT1G[0-9]+';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[[{'AT1G': 4}]]\n",
      "\n",
      "[[{'AT1G': 4}], [{'AT2G': 4}]]\n",
      "\n",
      "[[{'AT1G': 4}], [{'AT2G': 4}], [{'AT3G': 9}]]\n",
      "\n",
      "[[{'AT1G': 4}], [{'AT2G': 4}], [{'AT3G': 9}], [{'AT4G': 8}]]\n",
      "\n",
      "[[{'AT1G': 4}], [{'AT2G': 4}], [{'AT3G': 9}], [{'AT4G': 8}], [{'AT5G': 7}]]\n"
     ]
    }
   ],
   "source": [
    "Exercise3 = []\n",
    "for i in range(1,6):\n",
    "    Exercise3.append(crea_reportes(\"SELECT COUNT(*) AS {} FROM germplasm WHERE locus REGEXP '{}'\"\n",
    "                                   .format('AT' + str(i) + 'G', 'AT' + str(i) + 'G')))\n",
    "    print(Exercise3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id\tlocus\tgermplasm_code\tphenotype\tpubmed\tid_1\tgermplasm_id\tlocus\tgene\tprotein_length\n",
      "1\tAT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\t1\t1\tAT1G01040\tDCL1\t332\n",
      "2\tAT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\t2\t2\tAT1G01060\tLHY\t290\n",
      "3\tAT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\t3\t3\tAT1G01140\tCIPK9\t223\n",
      "4\tAT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\t4\t4\tAT1G01220\tFKGP\t190\n",
      "5\tAT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\t5\t5\tAT2G03720\tMRH6\t189\n",
      "6\tAT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\t6\t6\tAT2G03800\tGEK1\t196\n",
      "7\tAT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\t7\t7\tAT2G04240\tXERICO\t256\n",
      "8\tAT2G05210\tpot1-1\tNo visible phenotype.\t17627276\t8\t8\tAT2G05210\tPOT1A\t221\n",
      "9\tAT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\t9\t9\tAT3G02130\tRPK2\t284\n",
      "10\tAT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\t10\t10\tAT3G02140\tTMAC2\t300\n",
      "11\tAT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\t11\t11\tAT3G02230\tRGP1\t301\n",
      "12\tAT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\t12\t12\tAT3G02260\tBIG\t279\n",
      "13\tAT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\t13\t13\tAT3G02310\tSEP2\t175\n",
      "14\tAT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\t14\t14\tAT3G02680\tNBS1\t190\n",
      "15\tAT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\t15\t15\tAT3G02850\tSKOR\t234\n",
      "16\tAT3G02870\tvtc4-1\tascorbate deficient\t16595667\t16\t16\tAT3G02870\tVTC4\t311\n",
      "17\tAT3G03260\thdg8-1\tNo visible phenotype.\t16778018\t17\t17\tAT3G03260\tHDG8\t194\n",
      "18\tAT4G14790\tpdd17\tDefective pollen development.\t19237690\t18\t18\tAT4G14790\tSUV3\t312\n",
      "19\tAT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\t19\t19\tAT4G15210\tBAM5\t313\n",
      "20\tAT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\t20\t20\tAT4G15560\tDXS\t219\n",
      "21\tAT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\t21\t21\tAT4G15570\tMAA3\t294\n",
      "22\tAT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\t22\t22\tAT4G15802\tHSBP\t254\n",
      "23\tAT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\t23\t23\tAT4G15880\tESD4\t265\n",
      "24\tAT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\t24\t24\tAT4G16420\tADA2B\t279\n",
      "25\tAT4G16480\tatint4-2\tNo visible phenotype.\t16603666\t25\t25\tAT4G16480\tINT4\t284\n",
      "26\tAT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\t26\t26\tAT5G10480\tPAS2\t301\n",
      "27\tAT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\t27\t27\tAT5G10510\tAIL6\t310\n",
      "28\tAT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\t28\t28\tAT5G11110\tSPS2\t232\n",
      "29\tAT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\t29\t29\tAT5G11260\tHY5\t221\n",
      "30\tAT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\t30\t30\tAT5G11510\tMYB3R-4\t336\n",
      "31\tAT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\t31\t31\tAT5G12200\tPYD2\t310\n",
      "32\tAT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\t32\t32\tAT5G13290\tCRN\t189\n",
      "id\tlocus\tgermplasm_code\tphenotype\tpubmed\tid_1\tgermplasm_id\tlocus\tgene\tprotein_length\n",
      "15\tAT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\t15\t15\tAT3G02850\tSKOR\t234\n",
      "21\tAT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\t21\t21\tAT4G15570\tMAA3\t294\n",
      "[{'AT1G': 4}]\t[{'AT2G': 4}]\t[{'AT3G': 9}]\t[{'AT4G': 8}]\t[{'AT5G': 7}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import io\n",
    "\n",
    "with open('Resultados.tsv', 'w', newline='') as csvfile:\n",
    "    resultswriter = csv.writer(csvfile, delimiter='\\t', quotechar='\"')\n",
    "    resultswriter.writerow([\"id\", \"locus\", \"germplasm_code\", \"phenotype\", \"pubmed\", \"id_1\",\n",
    "                            \"germplasm_id\", \"locus\", \"gene\", \"protein_length\"])  \n",
    "    \n",
    "    for dictionary in Exercise1:\n",
    "        resultswriter.writerow([dictionary[\"id\"], dictionary[\"locus\"], dictionary[\"germplasm_code\"],\n",
    "                               dictionary[\"phenotype\"], dictionary[\"pubmed\"],dictionary[\"gene.id\"],\n",
    "                               dictionary[\"germplasm_id\"], dictionary[\"gene.locus\"], dictionary[\"gene_code\"],\n",
    "                               dictionary[\"protein_length\"]])\n",
    "        \n",
    "\n",
    "    \n",
    "    resultswriter.writerow([\"id\", \"locus\", \"germplasm_code\", \"phenotype\", \"pubmed\", \"id_1\",\n",
    "                            \"germplasm_id\", \"locus\", \"gene\", \"protein_length\"])\n",
    "    for dictionary in Exercise2:\n",
    "        resultswriter.writerow([dictionary[\"id\"], dictionary[\"locus\"], dictionary[\"germplasm_code\"],\n",
    "                               dictionary[\"phenotype\"], dictionary[\"pubmed\"],dictionary[\"gene.id\"],\n",
    "                               dictionary[\"germplasm_id\"], dictionary[\"gene.locus\"], dictionary[\"gene_code\"],\n",
    "                               dictionary[\"protein_length\"]])\n",
    "        \n",
    "        \n",
    "    resultswriter.writerow(Exercise3)\n",
    "    \n",
    "\n",
    "csvfile.close\n",
    "\n",
    "tsvfile = io.open(\"Resultados.tsv\")\n",
    "print(tsvfile.read())\n",
    "tsvfile.close()\n",
    "# for dictionary in Exercise1:\n",
    "#     print(dictionary)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\tAT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\t1\t1\tAT1G01040\tDCL1\t332\t\n",
      "2\tAT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\t2\t2\tAT1G01060\tLHY\t290\t\n",
      "3\tAT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\t3\t3\tAT1G01140\tCIPK9\t223\t\n",
      "4\tAT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\t4\t4\tAT1G01220\tFKGP\t190\t\n",
      "5\tAT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\t5\t5\tAT2G03720\tMRH6\t189\t\n",
      "6\tAT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\t6\t6\tAT2G03800\tGEK1\t196\t\n",
      "7\tAT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\t7\t7\tAT2G04240\tXERICO\t256\t\n",
      "8\tAT2G05210\tpot1-1\tNo visible phenotype.\t17627276\t8\t8\tAT2G05210\tPOT1A\t221\t\n",
      "9\tAT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\t9\t9\tAT3G02130\tRPK2\t284\t\n",
      "10\tAT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\t10\t10\tAT3G02140\tTMAC2\t300\t\n",
      "11\tAT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\t11\t11\tAT3G02230\tRGP1\t301\t\n",
      "12\tAT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\t12\t12\tAT3G02260\tBIG\t279\t\n",
      "13\tAT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\t13\t13\tAT3G02310\tSEP2\t175\t\n",
      "14\tAT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\t14\t14\tAT3G02680\tNBS1\t190\t\n",
      "15\tAT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\t15\t15\tAT3G02850\tSKOR\t234\t\n",
      "16\tAT3G02870\tvtc4-1\tascorbate deficient\t16595667\t16\t16\tAT3G02870\tVTC4\t311\t\n",
      "17\tAT3G03260\thdg8-1\tNo visible phenotype.\t16778018\t17\t17\tAT3G03260\tHDG8\t194\t\n",
      "18\tAT4G14790\tpdd17\tDefective pollen development.\t19237690\t18\t18\tAT4G14790\tSUV3\t312\t\n",
      "19\tAT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\t19\t19\tAT4G15210\tBAM5\t313\t\n",
      "20\tAT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\t20\t20\tAT4G15560\tDXS\t219\t\n",
      "21\tAT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\t21\t21\tAT4G15570\tMAA3\t294\t\n",
      "22\tAT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\t22\t22\tAT4G15802\tHSBP\t254\t\n",
      "23\tAT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\t23\t23\tAT4G15880\tESD4\t265\t\n",
      "24\tAT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\t24\t24\tAT4G16420\tADA2B\t279\t\n",
      "25\tAT4G16480\tatint4-2\tNo visible phenotype.\t16603666\t25\t25\tAT4G16480\tINT4\t284\t\n",
      "26\tAT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\t26\t26\tAT5G10480\tPAS2\t301\t\n",
      "27\tAT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\t27\t27\tAT5G10510\tAIL6\t310\t\n",
      "28\tAT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\t28\t28\tAT5G11110\tSPS2\t232\t\n",
      "29\tAT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\t29\t29\tAT5G11260\tHY5\t221\t\n",
      "30\tAT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\t30\t30\tAT5G11510\tMYB3R-4\t336\t\n",
      "31\tAT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\t31\t31\tAT5G12200\tPYD2\t310\t\n",
      "32\tAT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\t32\t32\tAT5G13290\tCRN\t189\t\n"
     ]
    }
   ],
   "source": [
    "for dictionary in Exercise1:\n",
    "    print(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t\".format(dictionary['id'], dictionary['locus'], dictionary['germplasm_code'], \n",
    "          dictionary['phenotype'], dictionary['pubmed'], dictionary['gene.id'],\n",
    "          dictionary['germplasm_id'], dictionary['gene.locus'], dictionary['gene_code'],\n",
    "          dictionary['protein_length']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
